packages/core/src/machines/authMachine.ts

```diff
@@ -0,0 +1,172 @@
+  signInActor,
+  signUpActor,
+  signOutActor,
+  resetPasswordActor,
+} from './actors';
+import { stopActor } from './actions';
+
+// TODO: Remove this before it's merged.
+if (typeof window !== 'undefined') {
+  inspect({
+    // options
+    // url: 'https://statecharts.io/inspect', // (default)
+    iframe: false, // open in new window
+  });
+}
+
+export const authMachine = createMachine<AuthContext, AuthEvent>(
+  {
+    id: 'auth',
+    initial: 'idle',
+    context: {},
+    states: {
+      // See: https://xstate.js.org/docs/guides/communication.html#invoking-promises
+      idle: {
+        invoke: [
+          {
+            // TODO Wait for Auth to be configured
+            src: 'getCurrentUser',
+            onDone: {
+              actions: 'setUser',
+              target: 'authenticated',
+            },
+            onError: 'signIn',
+          },
+          {
+            src: 'getAmplifyConfig',
+            onDone: {
+              actions: 'setAuthConfig',
+            },
+          },
+        ],
+      },
+      signIn: {
+        entry: 'spawnSignInActor',
+        exit: stopActor('signInActor'),
+        on: {
+          SIGN_UP: 'signUp',
+          RESET_PASSWORD: 'resetPassword',
+          'done.invoke.signInActor': [
+            {
+              target: 'signUp',
+              cond: 'shouldRedirectToSignUp',
+            },
+            {
+              target: 'authenticated',
+              actions: 'setUser',
+            },
+          ],
+        },
+      },
+      signUp: {
+        entry: 'spawnSignUpActor',
+        exit: stopActor('signUpActor'),
+        on: {
+          SIGN_IN: 'signIn',
+          'done.invoke.signUpActor': {
+            target: 'signIn',
+            actions: 'setUser',
+          },
+        },
+      },
+      resetPassword: {
+        entry: 'spawnResetPasswordActor',
+        exit: stopActor('resetPasswordActor'),
+        on: {
+          SIGN_IN: 'signIn',
+          'done.invoke.resetPasswordActor': 'signIn',
+        },
+      },
+      signOut: {
+        entry: 'spawnSignOutActor',
+        exit: [stopActor('signOutActor'), 'clearUser'],
+        on: { 'done.invoke.signOutActor': 'idle' },
+      },
+      authenticated: {
+        on: { SIGN_OUT: 'signOut' },
+      },
+    },
+    on: {
+      CHANGE: { actions: 'forwardToActor' },
+      SUBMIT: { actions: 'forwardToActor' },
+      FEDERATED_SIGN_IN: { actions: 'forwardToActor' },
+      RESEND: { actions: 'forwardToActor' },
+      SIGN_OUT: { actions: 'forwardToActor' },
+    },
+  },
+  {
+    actions: {
+      forwardToActor: forwardTo((context) => context.actorRef),
+      setUser: assign({
+        user: (_, event) => event.data.user || event.data,
+      }),
+      clearUser: assign({
+        user: undefined,
+      }),
+      setAuthConfig: assign({
+        config(_, event) {
+          return event.data.auth;
+        },
+      }),
+      spawnSignInActor: assign({
+        actorRef: (_, event) => {
```

Comment:
I'm curious:

**What are the pros/cons to spawning all machines when spawning the `authenticatorMachine`?**

```
const authenticatorMachine = createMachine({
  ...

  context: {
    // Optional, in case it's valuable to know the current screen (https://xstate.js.org/docs/tutorials/reddit.html#spawning-subreddit-actors)

    // Screens are already created (or as part of the `init` flow) and `idle`
	screen: signInMachine,
    signIn: signInMachine,
    signOut: signOutMachine,
  }
})
```

One thing that seems obvious is initializing `context` between screens: saving `username` on the Sign In screen and putting it on the `Sign Up` `context`.

Reply:
That's a great idea!! We can do what they did and still have `actorRef` active so that the references can be the same. This should also save us headache from starting and stopping actors.

One implementation detail I see is having to pass context after it's initialized, which is not a big deal but we would need some updateContext actions.